samza-core/src/main/java/org/apache/samza/processor/SamzaContainerController.java

```diff
@@ -107,7 +110,18 @@ public void startContainer(ContainerModel containerModel, Config config, int max
         Util.<String, MetricsReporter>javaMapAsScalaMap(metricsReporterMap),
         taskFactory);
     log.info("About to start container: " + containerModel.getContainerId());
-    containerFuture = executorService.submit(() -> container.run());
+    containerFuture = executorService.submit(() -> {
```

Comment:
>  Given that, I would suggest checking to see if startContainer has already been called here.

Yeah. You are right. Let me try to add in that check. 

> I would probably also consolidate stopContainer and shutdown into one method - shutdown really does mean shutdown with a single use controller.

Agreed. 

> I would either make them consistent (and e.g. have an awaitShutdown) or distinguish the behaviors through the method name (e.g. shutdownSync).

Let me consolidate stop and shutdown into one. Then, I can evaluate which option works better - shutdownSync or awaitShutdown. I don't have an opinion on this one way or the other. 

Thanks for the suggestions, @cpettitt  ! Very useful.